File: packages/java/engine/src/test/snapshot/examples/bst_insert_search.java
package com.algoflow.runner;

import com.algoflow.annotation.Tree;

public class Main {
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int val) { this.val = val; }
    }

    @Tree
    TreeNode root;

    public static void main(String[] args) {
        Main m = new Main();
        m.root = m.insert(m.root, 4);
        m.root = m.insert(m.root, 2);
        m.root = m.insert(m.root, 6);
        m.root = m.insert(m.root, 1);
        m.root = m.insert(m.root, 3);
        m.root = m.insert(m.root, 5);
        m.root = m.insert(m.root, 7);
        System.out.println("Found 3: " + m.search(m.root, 3));
        System.out.println("Found 8: " + m.search(m.root, 8));
    }

    public TreeNode insert(TreeNode node, int val) {
        if (node == null) return new TreeNode(val);
        if (val < node.val) node.left = insert(node.left, val);
        else if (val > node.val) node.right = insert(node.right, val);
        return node;
    }

    public boolean search(TreeNode node, int val) {
        if (node == null) return false;
        if (val == node.val) return true;
        return val < node.val ? search(node.left, val) : search(node.right, val);
    }
}

File: packages/python/engine/test/snapshot/examples/bst_insert_search.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def insert(node, val):
    if node is None:
        return TreeNode(val)
    if val < node.val:
        node.left = insert(node.left, val)
    elif val > node.val:
        node.right = insert(node.right, val)
    return node

def search(node, val):
    if node is None:
        return False
    if val == node.val:
        return True
    return search(node.left, val) if val < node.val else search(node.right, val)

root = None
for v in [4, 2, 6, 1, 3, 5, 7]:
    root = insert(root, v)
print("Found 3:", search(root, 3))
print("Found 8:", search(root, 8))
